### PR TITLE
Optimize the managed hash implementations.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR optimizes the managed hash implementations (`Sha1ForNonSecretPurposes` and SHA-2 which is used by browser) to reduce allocations and data copying. We make use of `InlineArray` to store the hash's state, and the SHA-2 classes were more significantly overhauled, to remove unsafe code, and support hashing the provided input directly without buffering it to a `MemoryStream`. Also on browser, one-shot hashing became allocation-free.

Let's see what CI thinks.

### CODE REVIEW

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderDispenser.Browser.cs [hunk 1]
```diff
@@ -53,9 +53,15 @@ public static class OneShotHashProvider
 
             public static int HashData(string hashAlgorithmId, ReadOnlySpan<byte> source, Span<byte> destination)
             {
-                HashProvider provider = CreateHashProvider(hashAlgorithmId);
-                provider.AppendHashData(source);
-                return provider.FinalizeHashAndReset(destination);
+                switch (hashAlgorithmId)
```
Comment (krwq) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderDispenser.Browser.cs:56]: is CreateHashProvider dead code now?

**Review by krwq — DISMISSED** [review]
Logic-wise looks good, would be good to get feedback from someone more familiar with inline arrays feature

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderDispenser.Browser.cs (hunk 1 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderDispenser.Browser.cs:56]: No, it is still being used by the non-one-shot hashes.

**Comment by krwq** [comment]
From my point of view assuming there is no issue with InlineArray this seems ok to merge. @GrabYourPitchforks @vcsjones have mentioned in the past they have some concerns with making changes to those classes and even suggested we should remove them if possible. We should wait for their feedback.

File: src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs [hunk 2]
```diff
@@ -155,5 +152,11 @@ private void Drain()
             _length += 512; // 64 bytes == 512 bits
             _pos = 0;
         }
+
+        [InlineArray(85)]
```
[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs:156]: No, will fix it.

[the remark this answers is not in the document]
Comment (bartonjs) [reply src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs:156]: In the discussion of `[InlineArray]` it was always stated that no one should be applying the attribute manually.  Did its companion language support not manifest, or should this be written in a different way?

[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs:156]: I don't think this one can be written without the attribute. I think the language can help with stuff like `Span<int> = [a, b, c];` but there is no syntax for stuff like `private int[184] x`.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs:156]: > Did its companion language support not manifest, or should this be written in a different way?

We ended up not shipping language syntax that would replace the need to apply [InlineArray] in these situations.  The compiler recognizes the attribute and let's index into such types and cast such types to spans, but applying [InlineArray(...)] is still the way to create such types when you need them in this fashion.

There are situations where the compiler will generate them for you, e.g. if you do:
```C#
Span<int> span = [1, 2, 3];
```
the compiler will ensure the assembly has an `[InlineArray(3)]` type that it'll use for this data.

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SHAHashProvider.Browser.Managed.cs [hunk 3]
```diff
@@ -154,94 +170,73 @@ private void InitializeState()
                 _stateSHA256[7] = 0x5be0cd19;
             }
 
+            public void Initialize()
+            {
+                this = new();
+            }
```
Comment (teo-tsirpanis) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SHAHashProvider.Browser.Managed.cs:176]: SP-800-108 tests are failing on browser, and my suspicion falls on this. Doesn't the default constructor above zero all members?

**Comment by jeffhandley** [comment]
Hi @teo-tsirpanis. You've clearly put a lot of thought and effort into this PR, and we appreciate your continued contributions. This particular PR causes us some concern though, so we won't be able to accept it.

CI is showing some failing unit tests, which of course is a blocker. It's also challenging for us to assess the value vs. risk of the substantial changes made to the browser implementations. As we noted in https://github.com/dotnet/designs/blob/main/accepted/2021/blazor-wasm-crypto.md, we are not aiming for these to be state-of-the-art implementations. As such, our tests mostly cover that we are calling an existing (assumed-to-be) correct implementation properly, not to show we implemented it correctly. Almost every change to an implementation is risky beyond what we can measure.

With the high risk and arguably low reward for changes here, we cannot proceed with the changes included in this PR. We would consider very small, targeted changes that address specific opportunities for improvement. The changes must be obviously and inarguably correct though. Examples could be replacing the shift-and-or with calling `ReadUInt32BigEndian`, where the correctness of that change would be clear. For the `Sha1ForNonSecretPurposes` changes, we're planning to replace this with a call to the system implementation, so it's not worth optimizing.

Thanks for your effort here and your understanding that we need to close this PR.

**Comment by jkotas** [comment]
@jeffhandley How about the changes outside the browser? Sha1ForNonSecretPurposes is used for non-security purposes for assembly name and tracing operations. It would be definitely nice to avoid the `byte[85]` array allocation there. I would be happy to sign-off on change that avoids the allocation in Sha1ForNonSecretPurposes.

> For the Sha1ForNonSecretPurposes changes, we're planning to replace this with a call to the system implementation, so it's not worth optimizing.

This was discussed multiple times before. IIRC, it is not viable to replace it with a call to system crypto libraries. We cannot break assembly names and tracing when the SHA1 is disallowed by system policy.

**Comment by GrabYourPitchforks** [comment]
> This was discussed multiple times before. IIRC, it is not viable to replace it with a call to system crypto libraries. We cannot break assembly names and tracing when the SHA1 is disallowed by system policy.

@jkotas Can you elaborate a bit on this? In .NET 5+, there's no block within the `SHA1` type which would forbid its instantiation by any caller. Such a block does (sadly) exist in .NET Framework, but I just looked at all the consumers of `Sha1ForNonSecretPurposes`, and it appears that all of them target netcore only. So I imagine everything should work just fine.

**Comment by jkotas** [comment]
Ok, I might have gotten the current blocking policies of legacy crypto algorithms in different environments mixed up. We have run into issues with SHA1 being [blocked in Redhat](https://github.com/dotnet/runtime/issues/65874) but it seems to be limited to signature verification only currently. The crypto library policies that block use of legacy crypto algorithms are only going to get stricter though.

In any case, introducing dependency on crypto libraries (openssl in particular) from assembly loader and runtime tracing would be a bad architecture decision. A simple `Console.WriteLine("Hello world")` should not have dependency on openssl. I think that the current `Sha1ForNonSecretPurposes` implementation is the right choice.

File: src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Sha1ForNonSecretPurposes.cs:156]: > will this work when targetting older versions of runtime?

This file isn't used downlevel, is it? I see it only being used in assemblies that are part of netcoreapp.

**Comment by teo-tsirpanis** [comment]
Thanks for the feedback @jeffhandley. I understand that having this PR combine quite a large number of changes in crypto code is hard to pass. Some of my changes fix significant performance problems (such as copying the entire input in a `MemoryStream` before hashing it) for which I will open a separate issue once I get some benchmarks running.

@jkotas sure, having SPC depend on OpenSSL is a bad idea and we can keep the managed implementation these, but on Windows and Apple platforms it wouldn't be a problem to depend on the system crypto libraries, would it?

**Comment by jkotas** [comment]
> Windows and Apple platforms it wouldn't be a problem to depend on the system crypto libraries, would it?

What would we gain by introducing the platform specific implementations? We prefer to have platform neutral implementations. `Sha1ForNonSecretPurposes` is used to hash very small amounts of data, and not very frequently.

**Comment by teo-tsirpanis** [comment]
> What would we gain by introducing the platform specific implementations?

Smaller binary size for one. #42537 tracks removing the managed SHA-1 implementations from where we can, let's move this discussion there.

**Pull request closed without merging**